affinity/cpp-20/d0796r3.md

```diff
@@ -589,17 +630,19 @@ The `affinity_query` class template provides an abstraction for a relative affin
 
 ## Free functions
 
-### `this_system::get_resources`
+### `this_system::discover_topology`
+
+The free function `this_system::discover_topology` is provided for discovering the **system topology**.
 
-The free function `this_system::get_resources` is provided for retrieving the `execution_resource`s which encapsulate the hardware platforms available within the system. We refer to these resources as the *system level resources*.
+	const execution_resource discover_topology();
 
-	std::vector<execution_resource> resources() noexcept;
+*Returns:* An `execution_resource` object exposing the **system execution resource**.
 
-*Returns:* An `std::vector` containing all *system level resources*.
+*Requires:* If `this_system::discover_topology().size() > 0`, `this_system::discover_topology()[0]` be the `execution_resource` use by `std::thread`. Calls to `this_system::discover_topology()` may not introduce a data race with any other call to `this_system::discover_topology()`.
 
-*Requires:* If `this_system::get_resources().size() > 0`, `this_system::get_resources()[0]` be the `execution_resource` use by `std::thread`. The value returned by `this_system::get_resources()` be the same at any point after the invocation of `main`.
+*Effects:* Discovers all **execution resources** available within the system and constructs the **system topology** DAG, describing a read-only snapshot at the point of the call.
 
-> [*Note:* Returning a `std::vector` allows users to potentially manipulate the container of `execution_resource`s after it is returned. We may want to replace this at a later date with an alternative type which is more restrictive, such as a range or span. *--end note*]
+*Throws:* Any exception thrown as a result of **system topology** discovery.
```

Comment:
So if this throws, does that mean the system topology is unusable? what can the user do to handle that? Either terminate or assume a sequential execution. Probably worth adding some questions for the committee as to what to do when the discovery fails, and what users would expect. Since the discover topology can call different libraries at different points in time, it may be that the first time is called works, but the second doesnt (if there is another _initialize_library_ in between). Even more interesting is the opposite, a call that fails, then a user disables the library, and then discover_topology works again. This should be described or clarified at some point (again, possibly not in this proposal!)

Reply:
Yes at the moment this is the case. This is something which we've discussed before, but I think we should spend some more time discussing. I responded on this in more detail in a comment above and created an issue for discussing it further (see https://github.com/codeplaysoftware/standards-proposals/issues/79).